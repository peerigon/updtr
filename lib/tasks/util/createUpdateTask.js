"use strict";

function createUpdateTask(instance, outdated) {
    return {
        name: outdated.name,
        type: outdated.type,
        updateTo: instance.wanted ? outdated.wanted : outdated.latest,
        rollbackTo: outdated.current,
    };
}

module.exports = createUpdateTask;
