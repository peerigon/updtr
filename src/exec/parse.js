const STRING_PROPERTIES = ["name", "current", "wanted", "latest"];

/**
 * Applies JSON.parse on the str.
 * Additionally, returns null if the str was empty or just contained whitespace.
 *
 * @param {string} str
 * @returns {*}
 */
function parse(str) {
    const trimmed = str.trim();

    if (trimmed.length === 0) {
        return null;
    }

    let resultJSON = {};
    try {
        resultJSON = JSON.parse(trimmed);
    } catch (e) {
        const eol = require("os").EOL;
        const lines = trimmed.split(eol);
        lines.map(stringJSON => {
            const possible = JSON.parse(stringJSON);
            if (possible.type === "table") {
                resultJSON = possible;
            }
        })
    }

    return resultJSON;
}

function returnIfValid(result) {
    STRING_PROPERTIES.forEach(prop => {
        if (typeof result[prop] !== "string") {
            throw new Error("Unexpected output format of package manager");
        }
    });

    return result;
}

function tryParse(normalizer) {
    return (stdout, cmd) => {
        try {
            return normalizer(parse(stdout));
        } catch (err) {
            err.message = `Error when trying to parse stdout from command '${ cmd }': ${ err.message }`;
            throw err;
        }
    };
}

function arrToObj(arr, keys) {
    return keys.reduce((obj, key, i) => {
        obj[key] = arr[i];

        return obj;
    }, {});
}

// By sorting the parsed data, we get deterministic results across different npm and yarn versions.
// As a nice side-effect, a package like eslint will always be updated before eslint-config-peerigon
// which might have a peer dependency on eslint
// See https://github.com/peerigon/updtr/issues/48
function sortByName(o1, o2) {
    return o1.name > o2.name;
}

const list = tryParse(
    parsed =>
        (parsed.dependencies === undefined ?
            [] :
            Object.keys(parsed.dependencies)
                .map(name => ({
                    name,
                    version: parsed.dependencies[name].version,
                }))
                .sort(sortByName))
);

export default {
    npm: {
        outdated: tryParse(parsed => {
            if (parsed === null) {
                return [];
            }

            const names = Object.keys(parsed);

            return names
                .map(name => parsed[name])
                .map((dep, index) =>
                    returnIfValid({
                        name: names[index],
                        current: dep.current,
                        wanted: dep.wanted,
                        latest: dep.latest,
                    })
                )
                .sort(sortByName);
        }),
        list,
    },
    yarn: {
        outdated: tryParse(
            parsed =>
                (parsed === null ?
                    [] :
                    parsed.data.body
                        .map(row => arrToObj(row, parsed.data.head))
                        .map(dep =>
                            returnIfValid({
                                name: dep.Package,
                                current: dep.Current,
                                wanted: dep.Wanted,
                                latest: dep.Latest,
                            })
                        )
                        .sort(sortByName))
        ),
        // We currently only use npm for the list command (see cmds.js).
        // Put the real implementation here if we decide to use yarn for list.
        list,
    },
};
