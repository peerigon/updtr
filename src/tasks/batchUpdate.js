import Sequence from "./util/Sequence";
import updateTo from "./util/updateTo";
import rollbackTo from "./util/rollbackTo";

function renderUpdate(updtr, updateTask) {
    return updtr.cmds.install({
        registry: updtr.config.registry,
        modules: updateTask.map(updateTo),
    });
}

function renderTest(updtr) {
    return updtr.cmds.test();
}

function renderRollback(updtr, failedUpdateTasks) {
    return updtr.cmds.install({
        registry: updtr.config.registry,
        modules: failedUpdateTasks.map(rollbackTo),
    });
}

async function update(sequence, updateTasks) {
    const updtr = sequence.updtr;
    let success;
    let testResult;

    await sequence.exec("updating", renderUpdate(updtr, updateTasks));
    try {
        testResult = await sequence.exec("testing", renderTest(updtr));
        success = true;
    } catch (err) {
        // Remember: instanceof Error might not work in Jest as expected
        // https://github.com/facebook/jest/issues/2549
        testResult = err;
        success = false;
    }

    sequence.baseEvent.success = success;

    if (success === false && updateTasks.length > 1) {
        // If the update was a failure, we roll back every update except the first.
        // The first update will be tested with the sequential-update. This way we skip one unnecessary install cycle.
        await sequence.exec(
            "rollback",
            renderRollback(updtr, updateTasks.slice(1))
        );
    }

    sequence.emit("result", {
        stdout: testResult.stdout,
    });

    return success;
}

export default (async function batchUpdate(updtr, updateTasks) {
    const sequence = new Sequence("batch-update", updtr, {
        updateTasks,
    });
    let success = true;

    if (updateTasks.length > 0) {
        sequence.start();
        success = await update(sequence, updateTasks);
        sequence.baseEvent = {
            success,
        };
        sequence.end();
    }

    return success;
});
