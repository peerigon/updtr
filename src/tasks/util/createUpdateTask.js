import {
    UPDATE_TO_LATEST,
    UPDATE_TO_NON_BREAKING,
    UPDATE_TO_WANTED,
} from "../../constants/config";

function determineUpdateToVersion({ current, wanted, latest }, { updateTo }) {
    switch (updateTo) {
        case UPDATE_TO_LATEST:
            return latest;
        case UPDATE_TO_WANTED:
            return wanted;
        case UPDATE_TO_NON_BREAKING:
        default:
            return "^" + current;
    }
}

export function isUpdateToNonBreaking(updateTask) {
    return updateTask.updateTo === "^" + updateTask.rollbackTo;
}

export default function createUpdateTask(outdated, updtrConfig) {
    return {
        name: outdated.name,
        updateTo: determineUpdateToVersion(outdated, updtrConfig),
        rollbackTo: outdated.current,
    };
}
