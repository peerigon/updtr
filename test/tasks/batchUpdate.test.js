import batchUpdate from "../../src/tasks/batchUpdate";
import FakeUpdtr from "../helpers/FakeUpdtr";
import createUpdateTask from "../../src/tasks/util/createUpdateTask";
import readFixtures from "../helpers/readFixtures";
import parse from "../../src/exec/parse";
import {
    execError,
    update,
    testPass,
    testFail,
    testFailWithRollback,
    errorExecUpdate,
    errorExecRollback,
} from "../fixtures/execResults";

let stdoutLogs;

function createUpdateTasks(updtrConfig) {
    const packageManager = updtrConfig.use;
    const outdated = parse[packageManager].outdated(
        stdoutLogs.get(`outdated/outdated.${ packageManager }.log`)
    );

    return outdated.map(outdated => createUpdateTask(outdated, updtrConfig));
}

beforeAll(async () => {
    stdoutLogs = await readFixtures([
        "outdated/outdated.npm.log",
        "outdated/outdated.yarn.log",
    ]);
});

describe("batchUpdate()", () => {
    describe("when the given updateTasks array is empty", () => {
        it("should resolve immediately with true without emitting events", async () => {
            const updtr = new FakeUpdtr();

            expect(await batchUpdate(updtr, [])).toBe(true);
            expect(updtr.exec.args).toEqual([]);
            expect(updtr.emit.args).toEqual([]);
        });
    });
    describe("when the given updateTasks array contains one update task", () => {
        describe("when the tests succeed", () => {
            it("should return true and execute expected commands and emit expected events", async () => {
                const updtr = new FakeUpdtr();
                const updateTasks = createUpdateTasks(updtr.config);

                updateTasks.length = 1;
                updtr.execResults = update.concat(testPass);

                const success = await batchUpdate(updtr, updateTasks);

                expect(success).toBe(true);
                expect(updtr.exec.args).toMatchSnapshot(
                    "one update task > test ok > exec args"
                );
                expect(updtr.emit.args).toMatchSnapshot(
                    "one update task > test ok > emit args"
                );
            });
        });
        describe("when the test fails", () => {
            it("should return false and execute expected commands and emit expected events", async () => {
                const updtr = new FakeUpdtr();
                const updateTasks = createUpdateTasks(updtr.config);

                updateTasks.length = 1;
                updtr.execResults = update.concat(testFail);

                const success = await batchUpdate(updtr, updateTasks);

                expect(success).toBe(false);
                expect(updtr.exec.args).toMatchSnapshot(
                    "one update task > test fail > exec args"
                );
                expect(updtr.emit.args).toMatchSnapshot(
                    "one update task > test fail > emit args"
                );
            });
        });
    });
    describe("when the given updateTasks array contains multiple update tasks", () => {
        describe("using npm", () => {
            describe("when the tests succeed", () => {
                it("should return true and execute expected commands and emit expected events", async () => {
                    const updtr = new FakeUpdtr();
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(testPass);

                    const success = await batchUpdate(updtr, updateTasks);

                    expect(success).toBe(true);
                    expect(updtr.exec.args).toMatchSnapshot(
                        "multiple update tasks > npm > test ok > exec args"
                    );
                    expect(updtr.emit.args).toMatchSnapshot(
                        "multiple update tasks > npm > test ok > emit args"
                    );
                });
            });
            describe("when the test fails", () => {
                it("should return false and execute expected commands and emit expected events", async () => {
                    const updtr = new FakeUpdtr();
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(testFailWithRollback);

                    const success = await batchUpdate(updtr, updateTasks);

                    expect(success).toBe(false);
                    expect(updtr.exec.args).toMatchSnapshot(
                        "multiple update tasks > npm > test fail > exec args"
                    );
                    expect(updtr.emit.args).toMatchSnapshot(
                        "multiple update tasks > npm > test fail > emit args"
                    );
                });
            });
        });
        describe("using yarn", () => {
            describe("when the tests succeed", () => {
                it("should return true and execute expected commands and emit expected events", async () => {
                    const updtr = new FakeUpdtr({
                        use: "yarn",
                    });
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(testPass);

                    const success = await batchUpdate(updtr, updateTasks);

                    expect(success).toBe(true);
                    expect(updtr.exec.args).toMatchSnapshot(
                        "multiple update tasks > yarn > test ok > exec args"
                    );
                    expect(updtr.emit.args).toMatchSnapshot(
                        "multiple update tasks > yarn > test ok > emit args"
                    );
                });
            });
            describe("when the test fails", () => {
                it("should return false and execute expected commands and emit expected events", async () => {
                    const updtr = new FakeUpdtr({
                        use: "yarn",
                    });
                    const updateTasks = createUpdateTasks(updtr.config);

                    updtr.execResults = update.concat(testFailWithRollback);

                    const success = await batchUpdate(updtr, updateTasks);

                    expect(success).toBe(false);
                    expect(updtr.exec.args).toMatchSnapshot(
                        "multiple update tasks > yarn > test ok > exec args"
                    );
                    expect(updtr.emit.args).toMatchSnapshot(
                        "multiple update tasks > yarn > test fail > emit args"
                    );
                });
            });
        });
    });
    describe("unexpected errors", () => {
        it("should completely bail out if the update cmd exits with a non-zero exit code", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            let givenErr;

            updtr.execResults = errorExecUpdate;

            try {
                await batchUpdate(updtr, updateTasks);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
            // emitted events: start, updating
            expect(updtr.emit.args.length).toBe(2);
        });
        it("should completely bail out if the rollback cmd exits with a non-zero exit code", async () => {
            const updtr = new FakeUpdtr();
            const updateTasks = createUpdateTasks(updtr.config);
            let givenErr;

            updtr.execResults = errorExecRollback;

            try {
                await batchUpdate(updtr, updateTasks);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr).toBe(execError);
            // emitted events: start, updating, testing, rollback
            expect(updtr.emit.args.length).toBe(4);
        });
    });
});
