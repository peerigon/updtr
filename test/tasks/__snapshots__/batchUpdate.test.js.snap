// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple update tasks > npm > test fail > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "npm test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/rollback",
    Object {
      "cmd": "npm install \\"updtr-test-module-2@2.0.0\\"",
      "success": false,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Oh noez",
      "success": false,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": false,
    },
  ],
]
`;

exports[`multiple update tasks > npm > test fail > exec args 1`] = `
Array [
  Array [
    "npm install \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "npm test",
  ],
  Array [
    "npm install \\"updtr-test-module-2@2.0.0\\"",
  ],
]
`;

exports[`multiple update tasks > npm > test ok > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "npm test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Everything ok",
      "success": true,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": true,
    },
  ],
]
`;

exports[`multiple update tasks > npm > test ok > exec args 1`] = `
Array [
  Array [
    "npm install \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "npm test",
  ],
]
`;

exports[`multiple update tasks > yarn > test fail > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "yarn add \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "yarn test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/rollback",
    Object {
      "cmd": "yarn add \\"updtr-test-module-2@2.0.0\\"",
      "success": false,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Oh noez",
      "success": false,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": false,
    },
  ],
]
`;

exports[`multiple update tasks > yarn > test ok > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "yarn add \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "yarn test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Everything ok",
      "success": true,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
        Object {
          "name": "updtr-test-module-2",
          "rollbackTo": "2.0.0",
          "updateTo": "2.1.1",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": true,
    },
  ],
]
`;

exports[`multiple update tasks > yarn > test ok > exec args 1`] = `
Array [
  Array [
    "yarn add \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "yarn test",
  ],
]
`;

exports[`multiple update tasks > yarn > test ok > exec args 2`] = `
Array [
  Array [
    "yarn add \\"updtr-test-module-1@2.0.0\\" \\"updtr-test-module-2@2.1.1\\"",
  ],
  Array [
    "yarn test",
  ],
  Array [
    "yarn add \\"updtr-test-module-2@2.0.0\\"",
  ],
]
`;

exports[`one update task > test fail > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@2.0.0\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "npm test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Oh noez",
      "success": false,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": false,
    },
  ],
]
`;

exports[`one update task > test fail > exec args 1`] = `
Array [
  Array [
    "npm install \\"updtr-test-module-1@2.0.0\\"",
  ],
  Array [
    "npm test",
  ],
]
`;

exports[`one update task > test ok > emit args 1`] = `
Array [
  Array [
    "batch-update/start",
    Object {
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/updating",
    Object {
      "cmd": "npm install \\"updtr-test-module-1@2.0.0\\"",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/testing",
    Object {
      "cmd": "npm test",
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/result",
    Object {
      "stdout": "Everything ok",
      "success": true,
      "updateTasks": Array [
        Object {
          "name": "updtr-test-module-1",
          "rollbackTo": "1.0.0",
          "updateTo": "2.0.0",
        },
      ],
    },
  ],
  Array [
    "batch-update/end",
    Object {
      "success": true,
    },
  ],
]
`;

exports[`one update task > test ok > exec args 1`] = `
Array [
  Array [
    "npm install \\"updtr-test-module-1@2.0.0\\"",
  ],
  Array [
    "npm test",
  ],
]
`;
